drivers/usb/device/usb_dc_usbip.c

```diff
@@ -0,0 +1,922 @@
+		/* TODO: continue read */
+		/* usbip_prep_rx(ep_idx, 0); */
+	}
+
+	return 0;
+}
+
+int usb_dc_ep_read(const uint8_t ep, uint8_t *const data,
+		   const uint32_t max_data_len, uint32_t * const read_bytes)
+{
+	LOG_DBG("ep 0x%02x max_data_len %u", ep, max_data_len);
+
+	if (usb_dc_ep_read_wait(ep, data, max_data_len, read_bytes) != 0) {
+		return -EINVAL;
+	}
+
+	if (!data && !max_data_len) {
+		/* When both buffer and max data to read are zero the above
+		 * call would fetch the data len and we simply return.
+		 */
+		return 0;
+	}
+
+	if (usb_dc_ep_read_continue(ep) != 0) {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+int usb_dc_ep_set_callback(const uint8_t ep, const usb_dc_ep_callback cb)
+{
+	uint8_t ep_idx = USB_EP_GET_IDX(ep);
+
+	LOG_DBG("Endpoint 0x%02x callback %p", ep, cb);
+
+	if (!usbip_ctrl.attached || !usbip_ep_is_valid(ep)) {
+		LOG_ERR("Not attached / wrong endpoint: 0x%x", ep);
+		return -EINVAL;
+	}
+
+	if (USB_EP_DIR_IS_IN(ep)) {
+		usbip_ctrl.in_ep_ctrl[ep_idx].cb = cb;
+	} else {
+		usbip_ctrl.out_ep_ctrl[ep_idx].cb = cb;
+	}
+
+	return 0;
+}
+
+void usb_dc_set_status_callback(const usb_dc_status_callback cb)
+{
+	usbip_ctrl.status_cb = cb;
+}
+
+int usb_dc_ep_mps(const uint8_t ep)
+{
+	uint8_t ep_idx = USB_EP_GET_IDX(ep);
+
+	if (!usbip_ctrl.attached || !usbip_ep_is_valid(ep)) {
+		LOG_ERR("Not attached / wrong endpoint: 0x%x", ep);
+		return -EINVAL;
+	}
+
+	if (USB_EP_DIR_IS_OUT(ep)) {
+		return usbip_ctrl.out_ep_ctrl[ep_idx].mps;
+	} else {
+		return usbip_ctrl.in_ep_ctrl[ep_idx].mps;
+	}
+}
+
+static int usbip_handle_control(struct usbip_command *cmd)
+{
+	uint8_t ep_idx = USB_EP_GET_IDX(cmd->hdr.ep);
+	struct usb_ep_ctrl_prv *ep_ctrl;
+
+	ep_ctrl = &usbip_ctrl.out_ep_ctrl[ep_idx];
+	if (ep_ctrl->cb == NULL) {
+		LOG_ERR("Control endpoint callback not set");
+		return -EIO;
+	}
+
+	if ((cmd->hdr.direction == USBIP_DIR_IN) ^
+	    (REQTYPE_GET_DIR(cmd->submit.bmRequestType) ==
+	     REQTYPE_DIR_TO_HOST)) {
+		LOG_ERR("Failed to verify bmRequestType");
+		return -EIO;
+	}
+
+	ep_ctrl->data_len = 8;
+	LOG_DBG("SETUP event ep 0x%02x %u", ep_idx, ep_ctrl->data_len);
+	memcpy(ep_ctrl->buf, &cmd->submit.bmRequestType, ep_ctrl->data_len);
+	ep_ctrl->cb(ep_idx, USB_DC_EP_SETUP);
+
+	if (cmd->hdr.direction == USBIP_DIR_OUT) {
+		/* Data OUT stage availably */
+		ep_ctrl->data_len = cmd->submit.length;
+		usbip_recv(ep_ctrl->buf, ep_ctrl->data_len);
+		LOG_DBG("DATA OUT event ep 0x%02x %u",
+			ep_idx, ep_ctrl->data_len);
+		ep_ctrl->cb(ep_idx, USB_DC_EP_DATA_OUT);
+	}
+
+	return 0;
+}
+
+static int usbip_handle_data(struct usbip_command *cmd)
+{
+	uint8_t ep_idx = cmd->hdr.ep;
+	struct usb_ep_ctrl_prv *ep_ctrl;
+	uint8_t ep;
+
+	if (cmd->hdr.direction == USBIP_DIR_OUT) {
+		if (ep_idx >= USBIP_OUT_EP_NUM) {
+			return -EINVAL;
+		}
+
+		ep_ctrl = &usbip_ctrl.out_ep_ctrl[ep_idx];
+		ep = ep_idx | USB_EP_DIR_OUT;
+		ep_ctrl->data_len = cmd->submit.length;
+		usbip_recv(ep_ctrl->buf, ep_ctrl->data_len);
```

Comment:
Same here, no error check?

Reply:
I have taken over this from usb_dc_native_posix_* for the time being. I am in the process of revising it.